### ISSUE TITLE
Unions without fields

### ISSUE THREAD

**Issue by ocharles** [dhall-lang/dhall-lang#224]
Currently if we want to have a union where only the tag is important, we write something like

```dhall
< Foo : {} | Bar : {} >
```

We can then use with `constructors` like

```
let alts = constructors < Foo : {} | Bar : {} > in alts.Foo {=}
```

However, having to supply this `{=}` is tedious, and adds nothing that couldn't be automatically inferred by Dhall itself. This gets even more frustrating if one uses `merge`, as all the handlers have to be functions `\(_ : {}) -> T`, which again is annoying as there is nothing I can do with this parameter. This can be slightly eased by defining `handler : T -> {} -> T`, but that still means I have to write more code and more annoying have to surround the actual merge handler in parenthesis.

Could Dhall support alternatives with no field at all? I would like to be able to write:

```dhall
let T = < Foo | Bar > in
let alts = constructors T in
let x = alts.Foo in
merge { Foo = True, Bar = False } x
```

**Comment by Gabriel439** [comment]
@ocharles: I'm reluctant to introduce this because it's only a constant-factor improvement to code complexity.  I prefer conceptual uniformity in the language over making the code more concise.

The only reason I exempted `Bool` and `Optional` from being modeled in the same way as other union types is because language bindings might want to give them special treatment (i.e. encode `Optional` as `null` or encode `Bool` using built-in language support for `bool`).

I'm less inclined to give enums special treatment because very few host languages that Dhall might bind to have built-in support for enums specifically.  They usually either encode enums using existing support for sum types (if the language is strongly typed) or integers/strings (if the language is weakly typed), and they rarely have a distinguished built-in support just for enums.  I can't think of any off the top of my head, but there might be some I haven't thought of.

**Comment by hanshoglund** [comment]
How about keeping the current semantics and introducing (optional) syntactic sugar for the empty case? E.g. 

```
< Foo | Bar >
===>
< Foo : {}, | Bar : {} >
```

And similarly for merge:

```
merge { Foo -> True, Bar -> False } x : T
===>
merge { Foo = \(_ : {}) -> True, Bar = \(_ : {}) -> False } x : T
```

This could even be mixed with the old syntax like so:

```
< None | Some : Natural >
===>
< None : {}, | Some : Natural >

merge { None -> 0, Some (x : Natural) -> x } y : T
===>
merge { None = \(_ : {}) -> 0, Some = \(x : Natural) -> x } y : T
```

**Comment by Gabriel439** [comment]
Thinking about this some more, I'm coming around to @ocharles's original proposal, which is to provide first-class support for empty union alternatives.

I'll restate the original proposal to make sure that I understood it correctly before I begin to standardize it (and also to reflect the recent changes to deprecate the `constructors` keyword).  My understanding is that of a user defines a union type of the form:

```haskell
Example = < Foo | Bar : {} | Baz : Text >
```

... then the types of the corresponding union constructors would be:

```haskell
Example.Foo : Example
Example.Bar : {} → Example
Example.Baz : Text → Example
```

... and the expected type of the record passed to `merge` would be:

```haskell
{ Foo : T
, Bar : {} → T
, Baz : Text → T
}
```

... such as:

```haskell
merge
{ Foo = "A"
, Bar = λ(_ : {}) → "B"
, Baz = λ(t : Text) → t
}
Example.Foo
```

This implies that the two alternatives named `Foo` and `Bar` would not be treated as equivalent, although I expect that users would never define an alternative with a field of type `{}` once this was supported.

If we go this route then at some point we might potentially go one step further and support this, too:

```haskell
Example = < Foo | Bar {} | Baz Natural Text >
```

In other words, support Haskell-style alternative syntax, which would imply these constructor types:

```haskell
Example.Foo : Example
Example.Bar : {} → Bar
Example.Baz : Natural → Text → Example
```

After all, it stands to reason that if you can support less than one argument to a union alternative, then you should also be able to support more than one argument, too.

However, for now I'll just focus on writing up a proposal to standardize empty union alternatives.

**Comment by sellout** [comment]
Yeah, I think I support this as part of the general Haskell-style “implicit product” approach. On its own, it seems a bit odd, I think.

@Gabriel439
> I expect that users would never define an alternative with a field of type `{}` once this was supported.

Maybe not explicitly, but it could easily come up as
`λ(a : Type) → < Foo | Bar : a | Baz : Text >` applied to `{}`.

**Comment by hanshoglund** [comment]
Note that [the 'more powerful union' solution](https://github.com/dhall-lang/dhall-lang/issues/224#issuecomment-451716152) increases the power of the "union" types to full ADTs, so you can now write 

```haskell
< Unit A >
  ~= { someKey : A }
  ~= { someOtherKey : A }
  ~= ...

< Pair Natural Text > 
  ~= { a : Natural, b : Text }
  ~= { b : Natural, c : Text } 
  ~= ...
```

This is a possible route to take, and @ocharles original proposal is indeed a limited version of this. 

I think I'd still prefer a purely syntactical approach to keep the type system simpler/more consistent.

**Comment by ocharles** [comment]
That's a good observation. For Dhall, I'm not sure I want un-named products
(at least at the syntactic level)

**Comment by Gabriel439** [comment]
@hanshoglund: Yes, this is part of the reason why I held off on empty alternatives for so long, because if you generalize them to support more than one field then they overlapped in functionality with products (i.e. anonymous records).

The thing that changed my mind on this was that even the "more powerful" Dhall unions would still be un-named products whereas records provide named products so there is still some distinction in functionality between the two.

Anyway, we don't have to add support anything other than empty alternatives for now.  I only brought up the more general union idea just to note that somebody will request that at some point so we will need to eventually explicitly accept or reject the idea.

**Comment by f-f** [comment]
Rereading the thread it looks like we all agree that Unions without fields (in the form specified by Gabriel [here](https://github.com/dhall-lang/dhall-lang/issues/224#issuecomment-451716152) - though not including the unnamed products extension yet) are a good thing to have, so I'll add the "standardize me" label to this issue

**Comment by Gabriel439** [comment]
I ran into one minor issue while attempting to standardize this.  There needs to be some way to distinguish between a union literal with a nullary constructor and a union type with a nullary constructor.

For example: is `< A >` a union type with a single nullary constructor or a union literal with a single nullary constructor?

My intuition is that we should optimize for the type being simpler and more readable, so I would prefer that union literals require some special syntax to distinguish them from the union type.  I'm not sure exactly what it should be, but it could be something like this

```haskell
< A * > : < A >

< A * | B > : < A | B >  -- Select the `A` nullary alternative

< A | B * > : < A | B > -- Select the `B` nullary alternative
```

Any other suggestions are welcome

**Comment by Profpatsch** [comment]
I’d keep it similar to how `{=} : {}` is handled, `<=A> : <A>`.

**Comment by f-f** [comment]
I'd agree with @Profpatsch's proposal, for consistency

**Comment by sellout** [comment]
Wouldn’t it be `<A=>`?

**Comment by ocharles** [comment]
I can't say I'm too fond of `<A=>`. `{=}` is ok because it's atomic syntax, but `<MySuperLongUnionType>` is very hard to visually distinguish from `<MySuperLongUnionType=>`

**Comment by Gabriel439** [comment]
@ocharles: I think a long alternative name is going to be hard to distinguish no matter which syntax we use

**Comment by Gabriel439** [comment]
One other thing to keep in mind is that it might help to pick a syntax for empty alternative literals that can generalize well to alternatives that can store anonymous products (should we choose to standardize that later on)

**Comment by ocharles** [comment]
Radical idea, but do you even want to have `<A | B * | C>` style syntax at the value level? This could just be `<A | B | C>.B`. To clarify, I'm literally suggesting dropping union-constructor syntax for values and having it only in the types.

**Comment by Gabriel439** [comment]
@ocharles: I like that idea, especially because it generalizes well to alternatives that are anonymous products

**Comment by Gabriel439** [comment]
Note that if we do this I might still support the old syntax as syntactic sugar for your proposed syntax to avoid disruption

**Comment by ocharles** [comment]
Yea, I just realised we only need this new syntax for nullary union constructors. The existing syntax can stay for unary ones.

**Comment by Gabriel439** [comment]
@ocharles: What I mean is that we can also use the new syntax for all union constructors (including non-nullary ones, too).  It's just that we support the old syntax for non-nullary constructors with a sufficiently long deprecation cycle to avoid disruption, along with `dhall lint` migration support, similar to how we handled `constructors`.

**Comment by Gabriel439** [comment]
Proposed change to the standard is up here: https://github.com/dhall-lang/dhall-lang/pull/438
